io-engine/src/rebuild/snapshot_rebuild.rs

```diff
@@ -0,0 +1,168 @@
+    ) -> Result<SnapshotRebuildJob, RebuildError> {
+        let url = super::parse_url(dst_uri)?;
+        let name = url.path().strip_prefix('/').unwrap_or(url.path());
+
+        if src_cr {
+            device_create(src_uri).await.context(SourceUriBdev)?;
+        }
+
+        match self.0.build(src_uri, dst_uri).await {
+            Ok(job) => Ok(SnapshotRebuildJob::new(name, src_cr, job)),
+            Err(error) => {
+                if src_cr {
+                    device_destroy(src_uri).await.ok();
+                }
+                Err(error)
+            }
+        }
+    }
+    /// Builds a `SnapshotRebuildJob` which can be started and which will then
+    /// rebuild from source uri to target local replica.
+    /// todo: probably target could still be a uri, example: lvol:///$uuid
+    ///  and then this would be handled the same way for non-replica targets.
+    pub async fn build(
+        self,
+        src_uri: &str,
+        replica_uuid: &str,
+    ) -> Result<SnapshotRebuildJob, RebuildError> {
+        // ensure that replica exists
+        // todo: when we have new backends, we can't just use `Lvol` directly.
+        let _lvol = Bdev::lookup_by_uuid_str(replica_uuid)
+            .ok_or(SnapshotRebuildError::ReplicaBdevNotFound {})
+            .and_then(|bdev| {
+                Lvol::try_from(bdev)
+                    .map_err(|_| SnapshotRebuildError::NotAReplica {})
+            })?;
+
+        device_create(src_uri).await.context(SourceUriBdev)?;
+
+        let dst_uri = format!("bdev:///{replica_uuid}");
+        match self.0.build(src_uri, &dst_uri).await {
+            Ok(job) => Ok(SnapshotRebuildJob::new(replica_uuid, true, job)),
+            Err(error) => {
+                device_destroy(src_uri).await.ok();
+                Err(error)
+            }
+        }
+    }
+}
+
+impl SnapshotRebuildJob {
+    /// Helps create a `Self` using a builder: `SnapshotRebuildJobBuilder`.
+    pub fn builder() -> SnapshotRebuildJobBuilder {
+        SnapshotRebuildJobBuilder::default().with_option(
+            RebuildJobOptions::default()
+                .with_read_opts(ReadOptions::CurrentUnwrittenFail),
+        )
+    }
+    fn new(name: &str, destroy_src_uri: bool, job: BdevRebuildJob) -> Self {
+        Self {
+            name: name.to_owned(),
+            inner: job,
+            destroy_src_uri,
+        }
+    }
+    /// Get a list of all snapshot rebuild jobs.
+    pub fn list() -> Vec<std::sync::Arc<SnapshotRebuildJob>> {
+        Self::get_instances().values().cloned().collect()
+    }
+    /// Get the name of this rebuild job.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+    /// Destroy this snapshot rebuild job itself.
+    pub fn destroy(self: std::sync::Arc<Self>) {
+        let _ = Self::remove(self.name());
+    }
+    /// Lookup a rebuild job by its name or target uri and return it.
+    pub fn lookup_any(name_or_uri: &str) -> Result<Arc<Self>, RebuildError> {
```

Comment:
I don't think that's a good idea to have the current lookup do this because you'd overload the lookup and there'd be no way of opting out of the uri check.
We could modify the existing lookup to take an enum, example:
```rust
lookup(Name("bob"))
lookup(NameOrUri("bob"))
```
Though seems to complicate things a bit since most lookups are using name anyway.